designs/nodepool-termination-grace-period.md

```diff
@@ -0,0 +1,175 @@
+* Cluster admins who want to allow users to protect long-running jobs from being interrupted by node disruption, up to a configured limit.
+
+This design adds in a new `terminationGracePeriod` field to the `NodePool` resource. This new field is mutable, and will impact any existing `NodeClaims` and `Nodes` associated with the `NodePool`.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  disruption:
+    terminationGracePeriod: 24h   # metav1.Duration, nil if not set
+```
+
+```
+$ kubectl explain nodepool.disruption.terminationGracePeriod
+KIND:     NodePool
+VERSION:  v1beta1
+
+FIELD:    terminationGracePeriod <metav1.Duration string>
+
+DESCRIPTION:
+  TerminationGracePeriod is the duration Karpenter will wait before forcefully terminating a node, measured from when
+  deletion is first initiated. Once the GracePeriod has expired, all pods on the node will be shutdown using the official
+  non-graceful shutdown taint.
+
+	Warning: this bypasses any PDB or terminationGracePeriodSeconds value set for a Pod.
+	Requires: K8s 1.26 or higher:  https://kubernetes.io/docs/concepts/architecture/nodes/#non-graceful-node-shutdown
+
+	This field is intended to be used by cluster administrators to enforce that nodes can be cycled within a given time period.
+	It can also be used to allow maximum time limits for long-running jobs which can delay node termination with preStop hooks.
+	If left undefined, Karpenter will wait indefinitely for pods to be drained.
+```
+
+## Code Definition
+
+```go
+type Disruption struct {
+  {...}
+  // terminationGracePeriod is a metav1.Duration, parsed as a duration string (15m, 24h, etc).
+  // The default value of nil means there is no timeout before pods will be forcibly evicted during node deletion.
+  terminationGracePeriod *metav1.Duration `json:"terminationGracePeriod"`
+}
+```
+
+## Alternative Spec
+Rather than defining the NodePool's `terminationGracePeriod` on the disruption spec, it could be added to the NodeClaim template instead.
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  template:
+    spec:
+      terminationGracePeriod: 24h   # metav1.Duration, nil if not set
+```
+
+#### Benefits
+1. Changes to the `terminationGracePeriod` would be less surprising, because current live nodes would not be impacted, the change would only take effect after new Nodes/NodeClaims had replaced old nodes.
+2. Behavior matches the behavior of a Pod's `terminationGracePeriodSeconds` more closely, which is not active until new pods have been created with a full pod rollout.
+
+#### Drawbacks
+1. Making updates to the `terminationGracePeriod` for a NodePool is cumbersome and disruptive to users of the cluster.
```

Comment:
I have a moderate preference to making this part of template, as it creates extensibility to standalone nodes. I agree that we should exclude it from drift, which is a slight deviation from how pods work, which is slightly surprising. However, both solutions deviate slightly in surprising ways, so I don't see the proposal and an advantage on this point. 

Making it part of template and excluding from drift is simply more powerful/extensible, without drawback compared to the proposal.

Reply:
>  I agree that we should exclude it from drift, which is a slight deviation from how pods work, which is slightly surprising

I have a few questions/comments that stem off the top-of-my-head if we choose to go the template route:
1. I do like that it's a standalone thing that just works if NodeClaims are generalized outside of NodePools. Up to this point, we have taken the stance that a NodeClaim can live on its own and serve as declarative state for a node launch and a node teardown. Having `terminationGracePeriod` in there is definitely nice
2. It raises a question in my head about what else is a stand-alone concept that we would put in there. `minLifetime`, `maxLifetime`? How different is `maxLifetime` from `expireAfter` today in the `disruption` block. For that matter, why is `expireAfter` in the disruption block if the concept can live on its own and we start putting other behavioral fields in the NodeClaim
3. The last one -- perhaps the one that I have the biggest question about -- if we don't drift on the value, what happens when someone updates the value in their NodePool that is different from the value in the NodeClaims? One option here is we reconcile the new value onto all of the owned NodeClaims. The downside of that approach is now how do you manually update the NodeClaim `terminationGracePeriod` if you want to shorten it in one-off cases. If the value is reconciled down, then you don't really have that option since whatever you change will just get reconciled back to whatever is in the NodePool. That bit feels awkward to me, since I'd like to think that the reason that we would choose to put it in the NodeClaim is so that manual interaction _is_ possible. The other options is you drift on it. You accept the painpoint _but_ you get the nice behavior that you can make manual changes to the NodeClaim and they won't be overridden after the NodeClaim's initial creation

Reply:
For 3, there's another option where we don't reconcile it down, and the controller just uses whatever happens to be in the nodeclaim. This may be "out of date" and a point in time value from the launch.

Reply:
That's true. I question, though, if that behavior is more surprising than just drifting to reconcile the state that the user requested in their NodePool.

Reply:
I could go either way on this one. At the end of the day, as long as we document the behavior, users will have to work with whether it drifts (or not). The value of having it on the NodeClaim seems clear.

Reply:
This is great because I have always wonder what is "NodeClaim" and what does it represent. The Karpenter docs doesn't have section on it. 

From my perspective I have always seen NodeClaim as just a Karpenter representation of a node. From that perspective I have always imagine all disruption related logic falls outside of NodeClaim as it is not an attribute of the node itself but rather the management of the node. Following that I would to say NodeClaims spec shouldn't have any information about disruption including maxLife or graceful node termination since these are actions taken not by the node itself but a service that manages the node. I wouldn't mind having annotations that have these information but thats purely for convenience

Reply:
> Following that I would to say NodeClaims spec shouldn't have any information about disruption including maxLife or graceful node termination since these are actions taken not by the node itself but a service that manages the node

I could see NodeClaim being a step above a "representation of the node". It's the node's manager IMO. I definitely think there's room to say that a NodeClaim is the declarative state for a Node, including providing details for the launch request as well as how the node gets spun down.

> since these are actions taken not by the node itself but a service that manages the node

The NodeClaim does orchestrate some of this logic today, at least for draining.

> The Karpenter docs doesn't have section on it.

I whole-heartedly agree that we should have a section in the docs. There's an issue that's open in the `aws/karpenter-provider-aws` repo for this, but this just isn't something that we have had the chance to get around to yet: https://github.com/aws/karpenter-provider-aws/issues/5144.

Reply:
> I could see NodeClaim being a step above a "representation of the node". It's the node's manager IMO. I definitely think there's room to say that a NodeClaim is the declarative state for a Node, including providing details for the launch request as well as how the node gets spun down.

I see. Thanks for the clarification.

I think its important to establish what is the responsibility of NodeClaims in order to know if NodeClaims should get disruption related information propagated to it. Agreed that I think it make sense that NodeClaim manages spin up and down however the current terminator controller logic seems to jump between interacting directly with the Node itself and NodeClaim's terminator controller which makes it very difficult to understand who is responsible.

From a tactical point of view, propagating termination related information down to NodeClaim is a one way door because it means NodeClaim should have the authority on how the termination happens. If we are undecided or need more time to determine if this is the right way, wouldn't it be safer to not do this right now? If we eventually decide that NodeClaim should have this responsibility, we should do it with fully conviction and completely migrate all termination logic into NodeClaim rather than partially doing now which continues building tech debt for little near term benefit.

Reply:
> is a one way door because it means NodeClaim should have the authority on how the termination happens

Agreed. This is something that we are thinking about for v1. It actually raises an even deeper question around whether expiration is a concept that should be stored strictly on the NodePool or if it also lives as a standalone concept on the NodeClaim. If you reframe the concept into `maxAge` or `maxLifetime`, all of a sudden you've got something that sounds like it might belong on the NodeClaim. I agree that if we take the step to push the `terminationGracePeriod` to the NodeClaim, we should do so with the plan to introduce expiration at the NodeClaim-level and drop it out of the NodePool.

Reply:
I'm in favor of putting terminationGracePeriod back into the NodeClaim spec simply to support users of NodeClaims without NodePools (I also think expiration should belong there too). I've updated the design to reflect that.